### ISSUE TITLE
Reason behind obfuscating define

### ISSUE THREAD

**Issue by fsgl1tch** [woodsmc/ostore#10]
What is the reason behind obfuscated defines like
```
#define START int retval = 0
#define IF_NOT_OK_HANDLE_ERROR( A ) if ( (retval = (A)) != ERR_OK ) HANDLE_ERROR( retval )
#define FINISH return retval;
```
I understand the need to abstract some library calls, but some of those defines seem to make it more difficult to follow the code.

**Comment by woodsmc** [comment]
That is a great question... it is to help with error handling. I've got a copy of this answer [available in the docs too](https://github.com/woodsmc/ostore/blob/master/docs/OStoreCodeErrorHandling.md).

# OStore Code Error Handling

Handling "c" code which calls functions in which each function returns a status code, or success code can be a pain. Ideally we'd check every single return / status code from every function we invoke. That's possible, but it tends to produce code which is a see of nested if statements.

In an effort to avoid this, the oStore code base uses a set of macros as follows:

```c
#define START int retval = 0
#define VALIDATE( A, B) if ( (A) ) HANDLE_ERROR(B)
#define IF_NOT_OK_HANDLE_ERROR( A ) if ( (retval = (A)) != ERR_OK ) HANDLE_ERROR( retval )
#define HANDLE_ERROR(A) { retval = A; goto HandleError; };
#define PROCESS_ERROR return retval; HandleError:
#define FINISH return retval;
```

These may seem weird at first, but they do allow us to write code which can handle errors a bit more gracefully:

```c
int exampleFunction(char** returnObj) {
    char* buffer = NULL;
    START;
    buffer = malloc(100);
    VALIDATE(buffer == NULL, NO_MEM);
    retval = callSomeFunction(buffer);
    IF_NOT_OK_HANDLE_ERROR(retval);
    (*returnObj) = buffer;
    
    PROCESS_ERROR;
    // something bad happened, clean up here
    if ( buffer ) {
        free(buffer);
    }
    (*returnObj) = NULL;
    FINISH;
}
```

If we expand the macro's you can see exactly what is going on.... 

```c
int exampleFunction(char** returnObj) {
    char* buffer = NULL;
    // START;
    int retval = 0; // 0 is ERR_OK 
    
    buffer = malloc(100);
    
    // VALIDATE(buffer == NULL, NO_MEM);
    if ( buffer == NULL ) { retval = NO_MEM; goto HandleError; }
    
    retval = callSomeFunction(buffer);
    //IF_NOT_OK_HANDLE_ERROR(retval);
    if ( retval != ERR_OK ) { goto HandleError; }
    
    // If we get this far, everything is ok, so return the buffer
    (*returnObj) = buffer;
    // the function will automatically return retval (ERR_OK).
    
    //PROCESS_ERROR;
    return retval;
    HandleError: // label <--
    // something bad happened, clean up here
    if ( buffer ) {
        free(buffer);
    }
    (*returnObj) = NULL;
    
    //FINISH;
    return retval;
}
```



As shown in the `ostore_create` function, this allows us to write code which can verify the return code of a function, and on an error jump to an error handling routine, before gracefully exiting the function call.

**Comment by fsgl1tch** [comment]
Hei,

Thank you for the detailed explanation, but aren't you trading nested if with macro expansion and goto? I think performance wise there is no big difference between nested if and the current solution. I would actually prefer the second version of the 2 codes posted, but this might just be preferences.

I would write the function like this though :)
```c
int exampleFunction(char** returnObj) 
{
    char* buffer = NULL;
    int retval = NO_MEM;
    
    buffer = malloc(100);
    if (buffer != NULL)
    {
        retval = callSomeFunction(buffer);

        if (retval == ERR_OK)
        {
            (*returnObj) = buffer;
        }
        else
        {
            free(buffer);
            (*returnObj) = NULL;
        }
    }

    return retval;
}
```

**Comment by woodsmc** [comment]
Hey there,

Yeah, your right there is no performance difference. It is ultimately a matter of choice, and I guess style. I've personally found that goto's and error handling really help where you are writing a function, which calls a lot of other functions, or which needs to check a lot of error conditions.  - In those circumstances the nested if statements quickly get super confusing.  If you check out `ostore_open` you will see what I mean, lots of condition checking as we open the file and try to validate the contents, this should be the permalink (see below). Not only does it get confusing, I didn't really fancy the RSI that comes with repeatedly trying `if (x) {...}` lol ;p  . Writing the native label and goto, each and every time is also a pain, and get's repetative very quick; hence the macros.

As an aside I think I saw similar styles used in some of Windows Terminal Team's code (it was presented in a youtube video I watched)... although, they use different macros, the basic approach of error handling, and jumping to the end / returning early with a clean up is the same.

https://github.com/woodsmc/ostore/blob/ba8a1552a17851fabff67c01faaaf071de5f063e/ostore/ostore.c#L39-L107
